services/placement/algo/mirrored.go

```diff
@@ -0,0 +1,299 @@
+}
+
+func (a mirroredAlgorithm) AddReplica(p services.Placement) (services.Placement, error) {
+	// TODO(cw): We could support AddReplica(p services.Placement, instances []services.PlacementInstance)
+	// and apply the shards from the new replica to the adding instances in the future.
+	return nil, errors.New("not supported")
+}
+
+func (a mirroredAlgorithm) RemoveInstances(
+	p services.Placement,
+	instanceIDs []string,
+) (services.Placement, error) {
+	if !p.IsMirrored() {
+		return nil, errMirrorAlgoOnNotMirroredPlacement
+	}
+
+	removingInstances := make([]services.PlacementInstance, len(instanceIDs))
+	for i, id := range instanceIDs {
+		instance, ok := p.Instance(id)
+		if !ok {
+			return nil, fmt.Errorf("instance %s not found in placement", id)
+		}
+		removingInstances[i] = instance
+	}
+
+	mirrorPlacement, err := mirrorFromPlacement(p)
+	if err != nil {
+		return nil, err
+	}
+	mirrorInstances, err := groupInstancesByShardSetID(removingInstances, p.ReplicaFactor())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, instance := range mirrorInstances {
+		ph, leavingInstance, err := newRemoveInstanceHelper(mirrorPlacement, instance.ID(), a.opts)
+		if err != nil {
+			return nil, err
+		}
+		// place the shards from the leaving instance to the rest of the cluster
+		if err := ph.PlaceShards(leavingInstance.Shards().All(), leavingInstance, ph.Instances()); err != nil {
+			return nil, err
+		}
+
+		if mirrorPlacement, _, err = addInstanceToPlacement(ph.GeneratePlacement(nonEmptyOnly), leavingInstance, false); err != nil {
+			return nil, err
+		}
+	}
+	return placementFromMirror(mirrorPlacement, nonLeavingInstances(p.Instances()), p.ReplicaFactor())
+}
+
+func (a mirroredAlgorithm) AddInstances(
+	p services.Placement,
+	addingInstances []services.PlacementInstance,
+) (services.Placement, error) {
+	if !p.IsMirrored() {
+		return nil, errMirrorAlgoOnNotMirroredPlacement
+	}
+
+	for _, instance := range addingInstances {
+		if _, exist := p.Instance(instance.ID()); exist {
+			// TODO(cw): Should we allow adding a leaving instance?
+			// What if leaving instance paired with new instances?
+			return nil, fmt.Errorf("instance %s already exist in the placement", instance.ID())
+		}
+	}
+
+	mirrorPlacement, err := mirrorFromPlacement(p)
+	if err != nil {
+		return nil, err
+	}
+	mirrorInstances, err := groupInstancesByShardSetID(addingInstances, p.ReplicaFactor())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, instance := range mirrorInstances {
+		if _, ok := mirrorPlacement.Instance(instance.ID()); ok {
+			return nil, fmt.Errorf("shard set id %s already exist in current placement", instance.ShardSetID())
+		}
+		ph := newAddInstanceHelper(mirrorPlacement, instance, a.opts)
+		ph.AddInstance(instance)
+		mirrorPlacement = ph.GeneratePlacement(nonEmptyOnly)
+	}
+
+	return placementFromMirror(mirrorPlacement, nonLeavingInstances(append(p.Instances(), addingInstances...)), p.ReplicaFactor())
```

Comment:
internally the placement contains a map of instances, p.Instances() will create a new slice each time so this should probably be fine.